src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubReceiver.java

```diff
@@ -0,0 +1,386 @@
+   * @return a new Subscriber Stub Instance
+   * @throws IOException the exception thrown by the Pub/Sub library if the Subscriber Stub Settings are invalid.
+   */
+  protected SubscriberStub getSubscriberStub(SubscriberStubSettings subscriberStubSettings) throws IOException {
+    return GrpcSubscriberStub.create(subscriberStubSettings);
+  }
+
+  /**
+   * Sleep for a given number of milliseconds, calculate new backoff time and return.
+   *
+   * @param backoff the time in milliseconds to delay execution.
+   * @return the new backoff delay in milliseconds
+   */
+  protected int sleepAndIncreaseBackoff(int backoff) {
+    try {
+      if (!isStopped()) {
+        Thread.sleep(backoff);
+      }
+    } catch (InterruptedException e) {
+      LOG.warn("Interrupted Exception in PubSubReceiver.");
+    }
+
+    return calculateUpdatedBackoff(backoff);
+  }
+
+  /**
+   * Calculate the updated backoff period baded on the Backoff configuration parameters.
+   *
+   * @param backoff the previous backoff period
+   * @return the updated backoff period for the next cycle.
+   */
+  protected int calculateUpdatedBackoff(int backoff) {
+    return Math.min((int) (backoff * backoffConfig.getBackoffFactor()), backoffConfig.getMaximumBackoffMs());
+  }
+
+  /**
+   * Get the rate at which this receiver should pull messages.
+   * <p>
+   * The default rate is 1000 (matching Apache Bahir) if the receiver has not been able to calculate a rate.
+   *
+   * @return The current rate at which this receiver should fetch messages.
+   */
+  protected int getMessageRate() {
+    int messageRate = supervisor().getCurrentRateLimit() < (long) Integer.MAX_VALUE ?
+      (int) supervisor().getCurrentRateLimit() : 1000;
+
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("Receiver rate is: " + messageRate);
+    }
+
+    return messageRate;
+  }
+
+  /**
+   * Class used to configure exponential backoff for Pub/Sub API requests.
+   */
+  public static class BackoffConfig implements Serializable {
+    final int initialBackoffMs;
+    final int maximumBackoffMs;
+    final double backoffFactor;
+
+    public BackoffConfig(int initialBackoffMs, int maximumBackoffMs, double backoffFactor) {
+      this.initialBackoffMs = initialBackoffMs;
+      this.maximumBackoffMs = maximumBackoffMs;
+      this.backoffFactor = backoffFactor;
+    }
+
+    public int getInitialBackoffMs() {
+      return initialBackoffMs;
+    }
+
+    public int getMaximumBackoffMs() {
+      return maximumBackoffMs;
+    }
+
+    public double getBackoffFactor() {
+      return backoffFactor;
+    }
+  }
+
+  /**
+   * Builder class for BackoffConfig
+   */
+  public static class BackoffConfigBuilder implements Serializable {
+    public int initialBackoffMs = 100;
+    public int maximumBackoffMs = 10000;
+    public double backoffFactor = 2.0;
+
+    protected BackoffConfigBuilder() {
+    }
+
+    public static BackoffConfigBuilder getInstance() {
```

Comment:
Why need a static method? Can't the caller just `new` it?

Reply:
Good point. I have noticed that most of Google's libraries use the this pattern for builders (see https://github.com/data-integrations/google-cloud/pull/425/commits/cfdbfff52860de20f372963f2217ced95a559687#diff-db0dd3323d1daa18e728d38181a3b88dacb74f0dbfcb9867d534c9ed730c5a9cR190), so my thinking was "Let's keep it consistent".

Reply:
I think it is better just new.